### PR TITLE
Fix warnings reported by MSVC (`-W2`) and clang-cl (`-W4`)

### PR DESCRIPTION
**Pull request by MisterDA**
This PR fixes warnings raised by MSVC at level 2, and clang-cl at level 4 (minus unused parameters and sign-compare). It should be rather uneventful. I think it's sane to enable _some_ warnings rather than none at all (the current), but C compilers (and sometimes programmers too) have a tendency of being overzealous when they report issues.

MSVC really likes to warn when an assignment of two different ranges, like float to int, is made without an explicit cast. If the warnings [C4244](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4244?view=msvc-170) and [C4146](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4146?view=msvc-170) are too annoying, they could also be disabled.

Clang-cl `-Wall` is equivalent to clang `-Weverything`, which is *really* too noisy, and enables some C++ warnings. Levels above 2 for MSVC are also noisy, and some warnings are even checking for C++ things in C. Someone with spare time might revisit this later…

### CODE REVIEW

**Review by nojb — COMMENTED** [review]
Thanks! A chunk of the diff is dead code removal and clearly correct. Some of the other changes involve introducing explicit type casts; it would be good to have a second opinion because this can sometimes be tricky. Otherwise, I am happy to approve this PR.

@dustanddreams maybe you can take a look? Thanks!

File: runtime/alloc.c
Comment (ghost) [inline runtime/alloc.c]: Why did you had to introduce a new local for the loop counter? `Enter_gc_preserve_vals` is only used from the `Do_alloc_small` macro next to it, which declares `mlsize_t i`.

Is there a warning about the use of `i` which is neither local to nor an argument of the macro? If so, such a too-pendantic warning ought to be disabled.

File: runtime/alloc.c
Comment (ghost) [inline runtime/alloc.c]: Almost the same question here. Why did you have to rename the variable and reduce its scope? Did you get a shadowing warning? If so, wouldn't simply rename from `i` to something else in both macros would be good enough?

File: runtime/ints.c
Comment (ghost) [inline runtime/ints.c]: Not really fond of this. Wouldn't `res = (uint64_t)-res` be good enough to silence those warnings?

File: runtime/major_gc.c
Comment (xavierleroy) [inline runtime/major_gc.c]: The original code is strange: why is `heap_words` a double-precision FP number when it obviously always holds integer values?  I'd rather declare `heap_words` as `uintnat` and fix the only place where it needs to be converted to FP:
```
  total_cycle_work =
    heap_sweep_words + (uintnat) ((double) heap_words * 100.0 / (100.0 + caml_percent_free));
```

File: runtime/ints.c
Comment (xavierleroy) [reply runtime/ints.c]: I agree with @dustanddreams 's comment.  Writing `~x + 1` instead of `-x` is correct but really hard to read.

File: runtime/caml/platform.h
Comment (xavierleroy) [inline runtime/caml/platform.h]: I'm used to "fetch and add", so "fetch and sub" feels a bit strange.  Why not declare the second parameter as `intnat v` and keep using "fetch and add -1" ?  This way the rest of the code doesn't need to be changed.

File: runtime/signals.c
Comment (xavierleroy) [inline runtime/signals.c:520]: Is the `(int)` cast necessary?  Why change the original code?

**Review by xavierleroy — COMMENTED** [review]
I reviewed this PR a while ago but apparently forgot to submit the review.  I'm globally in favor of this PR, but there are a few places where I think a better fix is warranted, see below.

File: runtime/alloc.c
Comment (MisterDA) [reply runtime/alloc.c]: `caml_alloc_9` defines parameters from `a` to `i`:
```c
CAMLexport value caml_alloc_9 (tag_t tag, value a, value b, value c, value d,
                               value e, value f, value g, value h, value i)
{
  Do_alloc_small(9, tag, a, b, c, d, e, f, g, h, i);
}
```
then `Do_alloc_small` expands to 
```c
#define Do_alloc_small(wosize, tag, ...)                \
{                                                       \
  Caml_check_caml_state();                              \
  value v;                                              \
  value vals[wosize] = {__VA_ARGS__};                   \
  mlsize_t i;                                           \
  CAMLassert ((tag) < 256);                             \
  // ...
```

the local `i` shadows the `i` parameter, thus the warning. Also, the `Enter_gc_preserve_vals` relies on the outer block declaring `i`, which I don't find particularly nice.
My intent was to kill two birds with one stone: first remove the shadowing warning by renaming the local `i` to `j`, then reduce the scope of the for loop iterators.

This was an incursion into MSVC level 4 warnings, so I'll just withdraw this commit.

File: runtime/alloc.c
Comment (MisterDA) [reply runtime/alloc.c]: See the answer above. I'll withdraw the commit.

File: runtime/caml/platform.h
Comment (MisterDA) [reply runtime/caml/platform.h]: It wouldn't work for direct calls to `atomic_fetch_sub`. There's just one of them, I'll keep it. Your idea works for the rest, thanks.

File: runtime/ints.c
Comment (MisterDA) [reply runtime/ints.c]: `res` is already `uint64_t`. Not sure how `(uint64_t)-(int64_t)res` really behaves.
`res = UINT64_MAX - res + 1` isn't an improvement.
What if we went with a comment? `res = ~res + 1 /* res = -res */`

File: runtime/major_gc.c
Comment (MisterDA) [reply runtime/major_gc.c]: Done, thanks.

File: runtime/signals.c
Comment (MisterDA) [reply runtime/signals.c:520]: I wanted to avoid a clang-cl warning, but I ended up disabling this warning (too much to "fix"). I also thought this change had a nice symmetry with the one above. 
```
runtime/signals.c(516,21): error: comparison of integers of different signs: 'int' and 'unsigned long long' [-Werror,-Wsign-compare]
  516 |   for (int i = 0; i < sizeof(posix_signals) / sizeof(int); i++)
      |                   ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: runtime/ints.c
Comment (ghost) [reply runtime/ints.c]: If `res` is already `int64_t`, then `(uint64_t)-res` will work nicely, since its value has already been bounds checked above so we know `0 <= res <= INT64_MAX`.

Comment (MisterDA) [reply runtime/ints.c]: I think you've misread, `res` is unsigned (`uint64_t`). The compiler warning is probably overzealous, but the compiler doesn't like when an unary minus is applied to an unsigned integer.

Comment (ghost) [reply runtime/ints.c]: Ah, so two casts. Ugly but still more human-friendly than bitmasking.

Comment (MisterDA) [reply runtime/ints.c]: Turns out a single cast is sufficient.

**Comment by MisterDA** [comment]
I think I've addressed all the comments, many thanks for the reviews.

**Comment by shindere** [comment]
@dustanddreams are you happy with the PR in its current state?

If so, I'm happy to approve on your behalf.

**Comment by MisterDA** [comment]
It's conflicting with trunk now, I need to rebase.

**Comment by shindere** [comment]
Antonin Décimo (2024/07/18 08:26 -0700):
> It's conflicting with trunk now, I need to rebase.

Oops, sorry, too quick.

**Comment by MisterDA** [comment]
Rebased!

**Comment by shindere** [comment]
Thanks. Happy to approve on somebody's behalf if necessary.

**Review by ghost — APPROVED** [review]
🚀

**Comment by MisterDA** [comment]
Rebased to fix conflicts.

**Comment by shindere** [comment]
Thanks. Approving this on behalf of @dustanddreams.

**Review by shindere — APPROVED** [review]

**Pull request merged**
